client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx

```diff
@@ -0,0 +1,165 @@
+import React, { useState } from 'react';
+import { ControlProps } from '@jsonforms/core';
+import {
+  useDebounceCallback,
+  useTranslation,
+  Box,
+  FormLabel,
+  Autocomplete,
+  IconButton,
+  EditIcon,
+  BasicTextInput,
+} from '@openmsupply-client/common';
+import {
+  FORM_LABEL_COLUMN_WIDTH,
+  FORM_INPUT_COLUMN_WIDTH,
+} from '../../common/styleConstants';
+import { z } from 'zod';
+import { useZodOptionsValidation } from '../../common/hooks/useZodOptionsValidation';
+import { useSearchQueries, QueryValues } from './useSearchQueries';
+
+const MIN_CHARS = 3;
+
+const Options = z.object({
+  /**
+   * Which pre-defined query to use (in useSearchQueries)
+   */
+  query: z.enum(QueryValues),
+  /**
+   * Pattern for formatting options list items (e.g. "${firstName} ${lastName}")
+   */
+  optionString: z.string().optional(),
+  /**
+   * Pattern for formatting selected result (as above)
+   */
+  displayString: z.string().optional(),
+  /**
+   * List of fields to save in document data (from selected item object)
+   */
+  saveFields: z.array(z.string()).optional(),
+  /**
+   * Text to show in input field before user entry
+   */
+  placeholderText: z.string().optional(),
+});
```

Comment:
Having a separate options validator in the child components seemed like a simpler option than a complex discriminated union at the top level.